### PR TITLE
Ascent: SoA Particle Support

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

Add support for pure SoA layouted particle containers for Ascent.

## Additional background

Follow-up to #2878.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by ax3l** [comment]
@atmyers @c-wetterer-nelson @cyrush @andrewcombs let's coordinate how we finalize this - this is the last element blocking the AMReX SoA transition for WarpX on our agenda :drum:

**Comment by c-wetterer-nelson** [comment]
> @atmyers @c-wetterer-nelson @cyrush let's coordinate how we finalize this - this is the last element blocking the AMReX SoA transition for WarpX on our agenda 🥁

RE: [the catalyst 2 support PR](https://github.com/AMReX-Codes/amrex/pull/3444). Yesterday was Andrew's last day so I will take over. The last piece there is a GitHub action script to add catalyst build testing to the CI. We could do that in a separate PR if you're comfortable with that. I am hoping to find time next week to make that happen. Otherwise, we are ready to review that PR.

File: Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H [hunk 1]
```diff
@@ -32,7 +31,7 @@ ParticleTileToBlueprint(const ParticleTile<amrex::Particle<NStructReal,
                         const std::string &topology_name)
 {
     int num_particles = ptile.GetArrayOfStructs().size();
-    int struct_size   = sizeof(Particle<NStructReal, NStructInt>);
+    int struct_size   = ParticleType::is_soa_particle ? 0 : sizeof(ParticleType);
```
Comment (cyrush) [inline Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H]: `struct_size` is used to represent element striding in the Conduit tree. 

For the SOA case - I expect we want native striding (for example, 8 bytes for float64), so I think we need to update the Particle Container to Conduit wrappers to handle these cases.

Comment (ax3l) [reply Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H]: Thank you Cyrus - where is that wrapper located? I thought it was this file...

File: Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H [hunk 2]
```diff
@@ -63,28 +59,55 @@ ParticleTileToBlueprint(const ParticleTile<ParticleType,
...
+                                          struct_size);
+#endif
+#if AMREX_SPACEDIM > 2
+        ParticleReal* zp = reinterpret_cast<ParticleReal*>(pbuf); pbuf += sizeof(ParticleReal);
+        n_coords["values/z"].set_external(zp,
+                                          num_particles,
+                                          0,
+                                          struct_size);
 #endif
+    }
```
Comment (cyrush) [inline Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H:110]: Yes, this is the style of logic we need.

If AOS, we stride by the struct size. If SOA, we stride by the native stride (set_external w/o the stride argument accomplishes this)

File: Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H [hunk 3]
```diff
@@ -94,65 +117,79 @@ ParticleTileToBlueprint(const ParticleTile<ParticleType,
     // -----------------------------
 
     int vname_real_idx = 0;
-    // struct real fields, the first set are always the particle positions
-    // which we wrap above
-    for (int i = 0; i < NStructReal; i++)
+    if constexpr(!ParticleType::is_soa_particle)
```
Comment (cyrush) [inline Src/Extern/Conduit/AMReX_Conduit_Blueprint_ParticlesI.H:120]: we need to build the fields for the SOA case, not sure how to access from AMReX but it should be a an else case.

**Review by cyrush — APPROVED** [review]

**Comment by ax3l** [comment]
Restarting CI.

From local tests, I think CI does not catch that his PR still has a compile error: #3639

**Comment by ax3l** [comment]
Based on `.github/workflows/ascent.yml` and #3639, testing locally with:
```
docker run -v $PWD:/opt/amrex -it alpinedav/ascent:0.9.2
```
```
export CXX=g++
export CC=gcc
export CMAKE_PREFIX_PATH=/ascent/install/lib/cmake/

. /ascent_docker_setup_env.sh

cd /opt/amrex/
cmake -S . -B build          \
           -DCMAKE_BUILD_TYPE=RelWithDebInfo \
           -DAMReX_ENABLE_TESTS=ON  \
           -DAMReX_FORTRAN=OFF      \
           -DAMReX_ASCENT=ON   \
           -DAMReX_CONDUIT=ON  \
           -DAMReX_PARTICLES=ON
cmake --build build -j 8
```

**Comment by ax3l** [comment]
I think our `main.cpp` is not yet complete :)

**Review by atmyers — APPROVED** [review]

**Pull request merged**